modulus/models/afno/modafno.py

```diff
@@ -0,0 +1,649 @@
+    ----------
+    in_features : int
+        Input feature size
+    latent_features : int
+        Latent feature size
+    out_features : int
+        Output feature size
+    activation_fn :  nn.Module, optional
+        Activation function, by default nn.GELU
+    drop : float, optional
+        Drop out rate, by default 0.0
+    scale_shift_kwargs : dict, optional
+        Options to the MLP that computes the scale-shift parameters
+    """
+
+    def __init__(
+        self,
+        in_features: int,
+        latent_features: int,
+        out_features: int,
+        mod_features: int,
+        activation_fn: nn.Module = nn.GELU(),
+        drop: float = 0.0,
+        scale_shift_kwargs: Union[dict, None] = None,
+    ):
+        super().__init__(
+            in_features=in_features,
+            latent_features=latent_features,
+            out_features=out_features,
+            activation_fn=activation_fn,
+            drop=drop,
+        )
+        if scale_shift_kwargs is None:
+            scale_shift_kwargs = {}
+        self.scale_shift = ScaleShiftMlp(
+            mod_features, latent_features, **scale_shift_kwargs
+        )
+
+    def forward(self, x: Tensor, mod_embed: Tensor) -> Tensor:
+        (scale, shift) = self.scale_shift(mod_embed)
+
+        scale_shift_shape = (scale.shape[0],) + (1,) * (x.ndim - 2) + (scale.shape[1],)
+        scale = scale.view(*scale_shift_shape)
+        shift = shift.view(*scale_shift_shape)
+
+        x = self.fc1(x)
+        x = x * scale + shift
+        x = self.act(x)
+        x = self.drop(x)
+        x = self.fc2(x)
+        x = self.drop(x)
+        return x
+
+
+class ModAFNO2DLayer(AFNO2DLayer):
+    """AFNO spectral convolution layer
+
+    Parameters
+    ----------
+    hidden_size : int
+        Feature dimensionality
+    mod_features : int
+        Number of modulation features
+    num_blocks : int, optional
+        Number of blocks used in the block diagonal weight matrix, by default 8
+    sparsity_threshold : float, optional
+        Sparsity threshold (softshrink) of spectral features, by default 0.01
+    hard_thresholding_fraction : float, optional
+        Threshold for limiting number of modes used [0,1], by default 1
+    hidden_size_factor : int, optional
+        Factor to increase spectral features by after weight multiplication, by default 1
+    scale_shift_kwargs : dict, optional
+        Options to the MLP that computes the scale-shift parameters
+    scale_shift_mode: str
+        If 'complex' (default), compute the scale-shift operation using complex
+        operations. If 'real', use real operations.
+    """
+
+    def __init__(
+        self,
+        hidden_size: int,
+        mod_features: int,
+        num_blocks: int = 8,
+        sparsity_threshold: float = 0.01,
+        hard_thresholding_fraction: float = 1,
+        hidden_size_factor: int = 1,
+        scale_shift_kwargs: Union[dict, None] = None,
+        scale_shift_mode: str = "complex",
```

Comment:
type `Literal["complex", "real"]` seems more appropriate here

Reply:
Great, will update - my knowledge of type annotations is rather incomplete.

Reply:
Done in https://github.com/NVIDIA/modulus/pull/547/commits/8e24b676645436ffe014e2a584a4c2404f429ec5